README.md
<h1 align="center">
	<img src="static/img/logo.png" style="height: 4em;" alt="VH7 Logo">
</h1>

<h3 align="center">
	A free and open source <u>URL shortening</u>, <u>file sharing</u> and <u>pastebin</u> service.
</h3>

<p align="center">
	<strong>
		<a href="https://vh7.uk/">Website</a>
	</strong>
</p>
<p align="center">
	<a href="https://ci.jakewalker.xyz/jake-walker/vh7"><img
        alt="Build Status"
    	src="https://img.shields.io/drone/build/jake-walker/vh7/master?server=https%3A%2F%2Fci.jakewalker.xyz&style=flat-square"></a>
</p>

<!-- TOC -->

- [Overview](#overview)
- [Getting Started](#getting-started)
    - [Deployment Notes](#deployment-notes)
    - [Docker](#docker)
        - [Tags](#tags)
        - [Manually Building Docker Image](#manually-building-docker-image)
- [Development](#development)
    - [Prerequisites](#prerequisites)
    - [Installation](#installation)
- [Running Tests](#running-tests)
    - [Code Style](#code-style)
- [Contributing](#contributing)
- [License](#license)

<!-- /TOC -->

## Overview

- **Free.** VH7 is not only free to use on the [official instance](https://vh7.uk) but is also free to download and run
for yourself.
- **Open Source.** All of VH7's source code is available here for the community to take a look under the hood. _We also
accept community contributions, just open a pull request!_
- **Multi-purpose.** Unlike other mainstream URL shorteners, VH7 also provides file sharing and a pastebin also with short
links.

## Getting Started

The recommended way of deploying is with [Docker](#docker), however, if you wish to deploy it with something else, there are instructions [here](http://docs.gunicorn.org/en/latest/deploy.html) on how to do so.

Check the [configuration page](https://github.com/jake-walker/vh7/wiki/Configuration) to see the different settings that VH7 can be configured with.

### Deployment Notes

- It is recommended to use a reverse proxy such as [Nginx](https://www.nginx.com/) or [Caddy](https://caddyserver.com/) between the internet and the instance of VH7.
- It is recommended to use a MySQL database instead of the default SQLite database.

### Docker

The suggested way of running VH7 is through Docker. Docker images are automatically built for every version of VH7 at [`jakewalker/vh7`](https://hub.docker.com/r/jakewalker/vh7).

```
docker volume create vh7_uploads
docker run --detach \
           --name vh7 \
           --restart always \
           -e VH7_DB_CONNECTION_STRING=mysql+mysqldb://username:password@hostname/database
           -e VH7_UPLOAD_FOLDER=/uploads
           -e VH7_SECRET=F6yZv5Xy8TBlYGkBs7P2wXlZqFAR3c
           -e VH7_SALT=wIkikFWxhAOV8O39B2JRCuyZqKQ8U1
           -e VH7_UPLOAD_MIN_AGE=30
           -e VH7_UPLOAD_MAX_AGE=90
           -e VH7_UPLOAD_MAX_SIZE=256
           -v vh7_uploads:/uploads
           -p 80:8000
           jakewalker/vh7:latest
```

#### Tags

The image [`jakewalker/vh7`](https://hub.docker.com/r/jakewalker/vh7) has the following tags:

- **`latest` → All** - Built on every commit to `master`, it should be stable to use in a production environment, but not as stable as the tags below.
- **`0.1`, `0.1.1` → Stable** - All version tags are more stable than `latest` and should be used where you would like the most stable version of VH7 as possible.

#### Manually Building Docker Image

```
docker build -t jakewalker/vh7 .
```

The built Docker image is saved as `vh7`. You can use then use the same command as above to run the newly built Docker image.

## Development

These instructions go over the method to setup a development environment for VH7. These instructions **should not** be used for a production setup of VH7.

### Prerequisites

- [Python 3.8](https://www.python.org/downloads/) (or any other version of Python 3)
- [Pipenv](https://github.com/pypa/pipenv) (for managing dependencies and a virtual environment)

### Installation

Clone the repository

```
git clone https://github.com/jake-walker/vh7
```

Install the dependencies

```
pipenv install --dev
```

Then setup the database with the latest schema

```
pipenv run flask db upgrade
```

Then finally, run

```
pipenv run python app.py
```

Open a browser to <https://localhost:5000/>.

## Running Tests

### Code Style

Code Style is checked by [flake8](http://flake8.pycqa.org/en/latest/). Simply run `flake8` inside the virtual environment to run the checks.

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please run tests and flake8 before committing.

## License

This project is licensed under the GNU General Public License v3.0 - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Image and video previews on uploaded files.
- Changelog page.

### Changed

- Bytecode and buffering is disabled in Dockerfile.
- Python 3.8 is now used by default instead of Python 3.7.
- Now using hCaptcha instead of Google reCAPTCHA.

### Fixed

- "SQL Server Gone Away" fix.
- Hide show API QR code button in account page if no API key is assigned to the account.

## [0.5.0] - 2020-04-25

### Added

- Google reCAPTCHA on login and registration pages.

### Fixed

- CORS headers are now correctly added to just API routes.


config.py
"""A file containing the configuration for the VH7 instance.

All of these values should be set through environment variables, but they can
be manually edited here if they need to be.
"""

import os
from distutils.dist import Distribution
from distutils.util import strtobool

dist = Distribution()
dist.parse_config_files()
VERSION = dist.get_option_dict("bumpversion")["current_version"][1]

# URI for creating a database connection with SQLAlchemy. Defaults to a local
# SQLite database.
SQLALCHEMY_DATABASE_URI = os.getenv("VH7_DB_CONNECTION_STRING",
                                    "sqlite:///db.sqlite")
SQLALCHEMY_TRACK_MODIFICATIONS = False
# Recycle connections after 1 hour
SQLALCHEMY_POOL_RECYCLE = 3600
SQLALCHEMY_POOL_PRE_PING = True

# The folder where uploaded files should go. Defaults to 'uploads' in the base
# of the project.
UPLOAD_FOLDER = os.getenv("VH7_UPLOAD_FOLDER", os.path.join(os.path.dirname(
                          os.path.realpath(__file__)), "uploads/"))

# A secret key for encrypting Flask cookies.
SECRET_KEY = os.getenv("VH7_SECRET", "keyboardcat")

# The alphabet that short links can be made from. Use only URL-safe characters
# and do not change after initial run!
HASHIDS_ALPHABET = ("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234"
                    "56789")

# The salt that short links are generated from. Do not change after initial
# run!
HASHIDS_SALT = os.getenv("VH7_SALT", "keyboardcat")

# The salt that is used for creating tokens (for example in confirmation
# emails)
TOKEN_SALT = os.getenv("VH7_TOKEN_SALT", "keyboardcat")

# The maximum age that tokens can be before expiring in seconds.
TOKEN_MAX_AGE = int(os.getenv("VH7_TOKEN_MAX_AGE", 7200))

# The minimum age that an uploaded file is stored for in days.
UPLOAD_MIN_AGE = int(os.getenv("VH7_UPLOAD_MIN_AGE", 30))
# The maximum age that an uploaded file is stored for in days.
UPLOAD_MAX_AGE = int(os.getenv("VH7_UPLOAD_MAX_AGE", 90))
# The maximum size that uploaded files are allowed to be in Mb.
UPLOAD_MAX_SIZE = int(os.getenv("VH7_UPLOAD_MAX_SIZE", 256))

# The maximum file size that the server will accept before aborting the
# request.
MAX_CONTENT_LENGTH = UPLOAD_MAX_SIZE * 1024 * 1024

# Enable flask-wtf CSRF tokens
WTF_CSRF_ENABLED = True

INSTANCE_NAME = os.getenv("VH7_INSTANCE_NAME", "Unnamed VH7 Instance")
INSTANCE_URL = os.getenv("VH7_INSTANCE_URL", "https://example.com/")
INSTANCE_OWNER = os.getenv("VH7_INSTANCE_OWNER", "Unknown")
INSTANCE_EMAIL = os.getenv("VH7_INSTANCE_EMAIL", "hi@example.com")
INSTANCE_COUNTRY = os.getenv("VH7_INSTANCE_COUNTRY", "United Kingdom")

# Mail configuration
MAIL_SERVER = os.getenv("VH7_MAIL_SERVER", "localhost")
MAIL_PORT = int(os.getenv("VH7_MAIL_PORT", 25))
MAIL_USE_TLS = strtobool(os.getenv("VH7_MAIL_TLS", "off"))
MAIL_USE_SSL = strtobool(os.getenv("VH7_MAIL_SSL", "off"))
MAIL_USERNAME = os.getenv("VH7_MAIL_USERNAME", None)
MAIL_PASSWORD = os.getenv("VH7_MAIL_PASSWORD", None)
MAIL_DEFAULT_SENDER = os.getenv("VH7_MAIL_SENDER", None)

# hCaptcha configuration
HCAPTCHA_SITE_KEY = os.getenv("VH7_HCAPTCHA_SITE", "")
HCAPTCHA_SECRET_KEY = os.getenv("VH7_HCAPTCHA_SECRET", "")
